_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+    
+    output <- self$linear(mlp_input)
+    
+    # shapes: (bs, 1) and (1, bs, hidden_size)
+    list(output, next_hidden)
+  }
+  
+)
+
+```
+
+### `seq2seq` module
+
+The `seq2seq` module is basically unchanged (apart from the fact that now, it allows for attention module configuration). For a detailed explanation of what happens here, please consult the [previous post](https://blogs.rstudio.com/ai/posts/2021-03-16-forecasting-time-series-with-torch_3/).
+
+```{r}
+seq2seq_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, attention_type, attention_size, n_forecast, 
+                        num_layers = 1, encoder_dropout = 0) {
+    
+    self$encoder <- encoder_module(type = type, input_size = input_size, hidden_size = hidden_size,
+                                   num_layers, encoder_dropout)
+    self$decoder <- decoder_module(type = type, input_size = 2 * hidden_size, hidden_size = hidden_size,
+                                   attention_type = attention_type, attention_size = attention_size, num_layers)
+    self$n_forecast <- n_forecast
+    
+  },
+  
+  forward = function(x, y, teacher_forcing_ratio) {
+    
+    outputs <- torch_zeros(dim(x)[1], self$n_forecast)$to(device = device)
+    encoded <- self$encoder(x)
+    encoder_outputs <- encoded[[1]]
+    hidden <- encoded[[2]]
+    # list of (batch_size, 1), (1, batch_size, hidden_size)
+    out <- self$decoder(x[ , n_timesteps, , drop = FALSE], hidden, encoder_outputs)
+    # (batch_size, 1)
+    pred <- out[[1]]
+    # (1, batch_size, hidden_size)
+    state <- out[[2]]
+    outputs[ , 1] <- pred$squeeze(2)
+    
+    for (t in 2:self$n_forecast) {
+      
+      teacher_forcing <- runif(1) < teacher_forcing_ratio
+      input <- if (teacher_forcing == TRUE) pred$unsqueeze(3) else y[ , t - 1]
+      out <- self$decoder(pred$unsqueeze(3), state, encoder_outputs)
+      pred <- out[[1]]
+      state <- out[[2]]
+      outputs[ , t] <- pred$squeeze(2)
+      
+    }
+    
+    outputs
+  }
+  
+)
+
+
+```
+
+When instantiating the top-level model, we now have an additional choice: that between additive and multiplicative attention. In the "accuracy" sense of performance, my tests did not show any differences; in the "speed" sense though, the multiplicative variant is a lot faster.
+
+```{r}
+net <- seq2seq_module("gru", input_size = 1, hidden_size = 32, attention_type = "multiplicative",
+                      attention_size = 8, n_forecast = n_timesteps)
+
+# training RNNs on the GPU currently prints a warning that may clutter 
+# the console
+# see https://github.com/mlverse/torch/issues/461
+# alternatively, use 
+# device <- "cpu"
+device <- torch_device(if (cuda_is_available()) "cuda" else "cpu")
+
+net <- net$to(device = device)
+```
+
+# Training
+
+Just like last time, in model training, we have the choice between degrees of teacher forcing. Below, we go with a fraction of 0.0, that is, no forcing at all.
```

Comment:
```suggestion
Just like last time, in model training, we choose the degree of teacher forcing. Below, we go with a fraction of 0.0, that is, no forcing at all.
```

Reply:
Technically speaking, "between" can only be applied to two options, and you use "among" for three or more. In this case, I think it's clearer to just say "choose," since it's numerical, and not a choice among options.

Reply:
oooh yes... crazy how you name things I just was vaguely uncomfortable with

btw I think I'll change to
> we get to choose the degree of teacher forcing

which still indicates the "agency thing" (???)